Throw `UnknownObjectException` for missing branch

jaseemabid opened the pull request:
Fixes #3011

Ideally this code should just check for status code and not touch the
message at all, but I'm sticking with the pattern here.

EnricoMi commented on github/Requester.py:
```diff
@@ -834,7 +834,11 @@ def createException(
-        elif status == 404 and (lc_message == "not found" or "no object found" in lc_message):
+        elif status == 404 and (
+            lc_message == "not found"  #
+            or lc_message == "branch not found"
```
what about
```suggestion
            or " not found" in lc_message
```

jaseemabid replied:
@EnricoMi Should I remove the check for message completely? This seems unnecessary if status code is 404. 

Can add tests 👍

EnricoMi reviewed (commented):
Can you add a test case that covers the particular error message so we have real record data?

jaseemabid commented:
@EnricoMi I've updated the PR and tests to reflect the new exception:

1. Replaced base `github.GithubException` with the more specific `github.UnknownObjectException`

2. Removed message check `self.assertEqual(raisedexp.exception.message, "Branch not found")` in tests

`createException()` doesn't pass a message to any specific exceptions, so this change is in line with other errors like `BadCredentialsException` or `RateLimitExceededException` which also don't have a message field set.

jaseemabid commented:
@EnricoMi Is this alright or do you need any additional changes?

jaseemabid commented:
@EnricoMi  Bump? 👋

EnricoMi commented:
Fixed in #3299.
